metagraph/src/annotation/row_diff_builder.cpp

```diff
@@ -214,32 +214,40 @@ void traverse_anno_chunked(
         before_chunk(block_size);
 
         succ_chunk.resize(block_size);
-        pred_chunk.resize(0);
-        pred_chunk_idx.resize(block_size + 1);
-        pred_chunk_idx[0] = 0;
-
         for (uint64_t i = 0; i < block_size; ++i) {
             succ_chunk[i] = succ[chunk + i];
-            pred_chunk_idx[i + 1] = pred_chunk_idx[i];
+        }
+
+        // read predecessor offsets
+        pred_chunk_idx.resize(block_size + 1);
+        pred_chunk_idx[0] = 0;
+        for (uint64_t i = 1; i <= block_size; ++i) {
+            // find where the last predecessor for the node ends
+            pred_chunk_idx[i] = pred_chunk_idx[i - 1];
             while (*pred_boundary_it == 0) {
-                ++pred_chunk_idx[i + 1];
-                pred_chunk.push_back(*pred_it);
-                ++pred_it;
+                ++pred_chunk_idx[i];
                 ++pred_boundary_it;
             }
             ++pred_boundary_it;
         }
 
+        // read all predecessors for the block
```

Comment:
Maybe add some runtime checks to make sure that the buffers are at the ends of the files? This could be a good way to check for corruption.

Reply:
This file is read partially (just a chunk).
Indeed, it would be good to add checks for correctness, but it doesn't seem straightforward.
Maybe we'll need to come up with a universal solution for such cases. Otherwise, it would require writing too many hacky checks, which is not good.